Objects/typeobject.c

```diff
@@ -3827,6 +3827,22 @@ type_new_impl(type_new_ctx *ctx)
     // Put the proper slots in place
     fixup_slot_dispatchers(type);
 
+    if (!_PyDict_HasOnlyStringKeys(type->tp_dict)) {
+        PyObject *name_of_class = type_name(type, NULL);
```

Comment:
Actually, all code in this file uses `type->tp_name` in error messages. So it would be better to use it here too. It will also make the code simpler.